src/main/kotlin/io/inventi/eventstore/projector/IdempotentEventHandler.kt

```diff
@@ -0,0 +1,175 @@
+package io.inventi.eventstore.projector
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.github.msemys.esjc.EventStore
+import com.github.msemys.esjc.EventStoreException
+import com.github.msemys.esjc.PersistentSubscription
+import com.github.msemys.esjc.PersistentSubscriptionCreateStatus
+import com.github.msemys.esjc.PersistentSubscriptionListener
+import com.github.msemys.esjc.PersistentSubscriptionSettings
+import com.github.msemys.esjc.RecordedEvent
+import com.github.msemys.esjc.RetryableResolvedEvent
+import com.github.msemys.esjc.SubscriptionDropReason
+import com.github.msemys.esjc.system.SystemConsumerStrategy
+import io.inventi.eventstore.projector.annotation.EventHandler
+import io.inventi.eventstore.projector.exception.UnsupportedMethodException
+import io.inventi.eventstore.projector.model.IdempotentEventClassifierRecord
+import io.inventi.eventstore.projector.model.MethodParametersType
+import io.inventi.eventstore.util.LoggerDelegate
+import org.jdbi.v3.core.statement.UnableToExecuteStatementException
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.context.SmartLifecycle
+import org.springframework.transaction.annotation.Transactional
+import java.lang.reflect.Method
+import java.util.concurrent.CompletionException
+
+
+abstract class IdempotentEventHandler(
+        private val streamName: String,
+        private val groupName: String
+) : SmartLifecycle {
+    companion object {
+        private val logger by LoggerDelegate()
+    }
+
+    @field:Value("\${eventstore.idempotent-event-classifier.table-name}")
+    lateinit var tableName: String
+
+    @Autowired
+    private lateinit var idempotentEventClassifierDao: IdempotentEventClassifierDao
+
+    @Autowired
+    private lateinit var eventStore: EventStore
+
+    @Autowired
+    private lateinit var objectMapper: ObjectMapper
+
+    var running: Boolean = false
+
+    override fun start() {
+        ensureSubscription()
+
+        val persistentSubscription = object : PersistentSubscriptionListener {
+
+            override fun onClose(subscription: PersistentSubscription?, reason: SubscriptionDropReason?, exception: Exception) {
+                if (exception is EventStoreException) {
+                    logger.warn("Reconnecting...")
+                    eventStore.subscribeToPersistent(streamName, groupName, this)
+                }
+
+                throw exception
+            }
+
+            @Transactional
```

Comment:
Spring doesn't know about this `PersistentSubscriptionListener`. Does `@Transactional` really work here?